File: src/main/java/io/github/intellij/dlanguage/stubs/types/IdentifierInitializerStubElementType.java
package io.github.intellij.dlanguage.stubs.types;

import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import io.github.intellij.dlanguage.psi.named.DLanguageIdentifierInitializer;
import io.github.intellij.dlanguage.psi.impl.named.DLanguageIdentifierInitializerImpl;
import io.github.intellij.dlanguage.resolve.processors.parameters.DAttributes;
import io.github.intellij.dlanguage.stubs.DLanguageIdentifierInitializerStub;
import java.io.IOException;
import org.jetbrains.annotations.NotNull;

public class IdentifierInitializerStubElementType extends
    DNamedStubElementType<DLanguageIdentifierInitializerStub, DLanguageIdentifierInitializer> {
    public IdentifierInitializerStubElementType(final String debugName) {
        super(debugName);
    }

    @Override
    public DLanguageIdentifierInitializer createPsi(@NotNull final DLanguageIdentifierInitializerStub stub) {
        return new DLanguageIdentifierInitializerImpl(stub, this);
    }

    @NotNull
    @Override
    public DLanguageIdentifierInitializerStub createStub(@NotNull final DLanguageIdentifierInitializer psi,
                                                         final StubElement parentStub) {
        return new DLanguageIdentifierInitializerStub(parentStub, this, psi.getName(), psi.getAttributes());
    }

    @Override
    public void serialize(@NotNull final DLanguageIdentifierInitializerStub stub, @NotNull final StubOutputStream dataStream) throws IOException {
        dataStream.writeName(stub.getName());
        stub.getAttributes().write(dataStream);
    }

    @NotNull
    @Override
    public DLanguageIdentifierInitializerStub deserialize(@NotNull final StubInputStream dataStream, final StubElement parentStub) throws IOException {
        return new DLanguageIdentifierInitializerStub(parentStub, this, dataStream.readName(),
            DAttributes.Companion.read(dataStream));
    }
}

File: dlang/psi-impl/src/main/kotlin/io/github/intellij/dlanguage/stubs/types/IdentifierInitializerStubElementType.kt
package io.github.intellij.dlanguage.stubs.types

import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import io.github.intellij.dlanguage.psi.impl.named.DLanguageIdentifierInitializerImpl
import io.github.intellij.dlanguage.psi.named.DLanguageIdentifierInitializer
import io.github.intellij.dlanguage.resolve.processors.parameters.DAttributes.Companion.read
import io.github.intellij.dlanguage.stubs.DLanguageIdentifierInitializerStub
import java.io.IOException

class IdentifierInitializerStubElementType(debugName: String) :
    DNamedStubElementType<DLanguageIdentifierInitializerStub, DLanguageIdentifierInitializer>(debugName) {
    override fun createPsi(stub: DLanguageIdentifierInitializerStub): DLanguageIdentifierInitializer {
        return DLanguageIdentifierInitializerImpl(stub, this)
    }

    override fun createStub(
        psi: DLanguageIdentifierInitializer,
        parentStub: StubElement<*>?
    ): DLanguageIdentifierInitializerStub {
        return DLanguageIdentifierInitializerStub(parentStub, this, psi.name, psi.attributes)
    }

    @Throws(IOException::class)
    override fun serialize(stub: DLanguageIdentifierInitializerStub, dataStream: StubOutputStream) {
        dataStream.writeName(stub.name)
        stub.attributes.write(dataStream)
    }

    @Throws(IOException::class)
    override fun deserialize(
        dataStream: StubInputStream,
        parentStub: StubElement<*>?
    ): DLanguageIdentifierInitializerStub {
        return DLanguageIdentifierInitializerStub(
            parentStub, this, dataStream.readName(),
            read(dataStream)
        )
    }
}
